test/integration_test/basic_test.go

```diff
@@ -6,186 +6,206 @@ import (
+					NodeAffinity: &v1.NodeAffinity{
+						RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+							NodeSelectorTerms: []v1.NodeSelectorTerm{
+								{
+									MatchExpressions: []v1.NodeSelectorRequirement{
+										{
+											Key:      labelKeySkipPX,
+											Operator: v1.NodeSelectorOpNotIn,
+											Values:   []string{labelValueTrue},
+										},
+									},
+								},
 							},
 						},
 					},
 				},
 			},
 		},
-	}
+		ShouldSkip: false,
+		TestFunc:   BasicInstallWithNodeAffinity,
+	},
+	{
+		TestName:        "Upgrade",
+		TestrailCaseIDs: []string{},
+		TestStorageCluster: &op_corev1.StorageCluster{
+			ObjectMeta: meta.ObjectMeta{Name: "upgrade-test"},
+		},
+		ShouldSkip: false,
+		TestFunc:   BasicUpgrade,
+	},
+}
 
-	// Get K8S nodes
-	nodeList, err := coreops.Instance().GetNodes()
-	require.NoError(t, err)
+func TestStorageClusterBasic(t *testing.T) {
+	for _, testCase := range testStorageClusterBasicCases {
+		testCase.RunTest(t)
+	}
+}
 
-	// Set Node Affinity label one of the K8S nodes
-	var nodeNameWithLabel string
-	for _, node := range nodeList.Items {
-		if coreops.Instance().IsNodeMaster(node) {
-			continue // Skip master node, we don't need to label it
+func BasicInstall(tc *TestCase) func(*testing.T) {
+	return func(t *testing.T) {
+		if tc.ShouldSkip {
+			t.Skip()
 		}
-		logrus.Infof("Label node %s with %s=%s", node.Name, labelKey, labelValue)
-		if err := coreops.Instance().AddLabelOnNode(node.Name, labelKey, labelValue); err != nil {
-			require.NoError(t, err)
-		}
-		nodeNameWithLabel = node.Name
-		break
-	}
 
-	// Deploy cluster
-	logrus.Infof("Create StorageCluster %s in %s", cluster.Name, cluster.Namespace)
-	cluster, err = createStorageCluster(cluster)
-	require.NoError(t, err)
-
-	// Validate cluster deployment
-	logrus.Infof("Validate StorageCluster %s", cluster.Name)
-	err = testutil.ValidateStorageCluster(imageListMap, cluster, defaultValidateDeployTimeout, defaultValidateDeployRetryInterval, true, "")
-	require.NoError(t, err)
-
-	// Delete cluster
-	logrus.Infof("Delete StorageCluster %s", cluster.Name)
-	err = testutil.UninstallStorageCluster(cluster)
-	require.NoError(t, err)
-
-	// Validate cluster deletion
-	logrus.Infof("Validate StorageCluster %s deletion", cluster.Name)
-	err = testutil.ValidateUninstallStorageCluster(cluster, defaultValidateUninstallTimeout, defaultValidateUninstallRetryInterval)
-	require.NoError(t, err)
-
-	// Remove Node Affinity label from the node
-	logrus.Infof("Remove label %s from node %s", nodeNameWithLabel, labelKey)
-	if err := coreops.Instance().RemoveLabelOnNode(nodeNameWithLabel, labelKey); err != nil {
+		// Get versions from URL
+		logrus.Infof("Get component images from versions URL")
+		imageListMap, err := testutil.GetImagesFromVersionURL(pxSpecGenURL)
+		require.NoError(t, err)
+
+		// Construct Portworx StorageCluster object
+		err = constructStorageClusterForTestCase(tc, pxSpecGenURL, imageListMap)
```

Comment:
The storage cluster should be input, if we create it inside of the test, it's more like internal variable, there is no point to save it back to TC.  

As we discussed in last friday's meeting, I think storagecluster is not a common input for all test cases and it should not be in TestCase struct. For examples:
- for Daemonset to operator migration test cases, the input should be Daemonset and not StorageCluster.
- If we implement autopilot functional tests, not installation of autopilot via operator, autopilot project functional tests. We will have installation of autopilot covered by operator, and functionality of autopilot would be tested via a bunch of test cases that assumes autopilot is already installed, this struct would not fit.

I think a common test case consists of: input, output, setup, execution and teardown, which are common parameters.

Reply:
@haibinxie I thought we decided that we can keep the storagecluster in the case struct, but it should be optional. That way, we can use the same test func to test similar test cases by only changing the input storagecluster. 
The tests that don't need or don't want to have an input StorageCluster will keep it nil in their struct and create their own in the test func.

Reply:
>The storage cluster should be input, if we create it inside of the test, it's more like internal variable, there is no point to save it back to TC.

we have the storage cluster as test case input, the construct here only populate the env var and images. the constructed stc is still belongs to the current TC so there's actually no difference this is more like a personal preference, I planned to make the function a call belongs to TC object.
>As we discussed in last friday's meeting, I think storagecluster is not a common input for all test cases and it should not be in TestCase struct.

what about we make the field `InputSpec interface{}`? then we could assert the type in different test functions, that way it becomes more scalable for different scenarios?  @haibinxie 
also, the idea of setup a TestFunc field would be applicable for other steps like setup and teardown func as an input spec if we would like to do so, I feel it's feasible.

Reply:
Yes, making it a InputSpec sounds good, and different test can have different input as desired.

Reply:
When we look at the code, the input StorageCluster is empty with just a name, then it gets changed inside of the function, which loses the goal of making StorageCluster as a configurable input and looks confusing. 

`we have the storage cluster as test case input, the construct here only populate the env var and images. the constructed stc is still belongs to the current TC so there's actually no difference this is more like a personal preference, I planned to make the function a call belongs to TC object.

`

Reply:
got it, what about this, let's make the InputSpec a `func() interface` similar to ShouldSkip, TestFunc, this way it's more clear, could be defined in testcase list, also fully customizable, and it's consistent with other fields.

Reply:
sounds good to me. @piyush-nimbalkar what do you think?

Reply:
Sounds good, but honestly I don't see any other input spec coming up other than StorageCluster in the near term. This could be just a storagecluster spec or if you want to keep it function it could just return the StorageCluster spec. We can extend this in future when we have different kinds of inputs. Also, for migration it won't be just one spec but a bunch of specs. So, it would be better to create those in the test itself.

I agree with Michael, that we should probably not create an input spec if it is not passed. Let the individual test be responsible for that.